src/Simulation/TargetDefinitions/Decompositions/Measure.qs

```diff
@@ -40,29 +41,115 @@ namespace Microsoft.Quantum.Intrinsic {
+            }
+            return MResetZ(q);
+        }
+        elif Length(bases) == 5 {
+            use q = Qubit();
+            within {
+                H(q);
+            }
+            apply {
+                JointMeasureHelper(bases[0], q, qubits[0]);
+                JointMeasureHelper(bases[1], q, qubits[1]);
+                JointMeasureHelper(bases[2], q, qubits[2]);
+                JointMeasureHelper(bases[3], q, qubits[3]);
+                JointMeasureHelper(bases[4], q, qubits[4]);
+            }
+            return MResetZ(q);
+        }
+        elif Length(bases) == 6 {
+            use q = Qubit();
+            within {
                 H(q);
-                for k in 0 .. Length(bases) - 1 {
-                    if( bases[k] == PauliX ) { Controlled X ([q], qubits[k]); }
-                    if( bases[k] == PauliZ ) { Controlled Z ([q], qubits[k]); }
-                    if( bases[k] == PauliY ) { Controlled Y ([q], qubits[k]); }
-                }
+            }
+            apply {
+                JointMeasureHelper(bases[0], q, qubits[0]);
+                JointMeasureHelper(bases[1], q, qubits[1]);
+                JointMeasureHelper(bases[2], q, qubits[2]);
+                JointMeasureHelper(bases[3], q, qubits[3]);
+                JointMeasureHelper(bases[4], q, qubits[4]);
+                JointMeasureHelper(bases[5], q, qubits[5]);
+            }
+            return MResetZ(q);
+        }
+        elif Length(bases) == 7 {
+            use q = Qubit();
+            within {
                 H(q);
-                set res =  M(q);
-                Reset(q);
             }
+            apply {
+                JointMeasureHelper(bases[0], q, qubits[0]);
+                JointMeasureHelper(bases[1], q, qubits[1]);
+                JointMeasureHelper(bases[2], q, qubits[2]);
+                JointMeasureHelper(bases[3], q, qubits[3]);
+                JointMeasureHelper(bases[4], q, qubits[4]);
+                JointMeasureHelper(bases[5], q, qubits[5]);
+                JointMeasureHelper(bases[6], q, qubits[6]);
+            }
+            return MResetZ(q);
+        }
+        elif Length(bases) == 8 {
+            use q = Qubit();
+            within {
+                H(q);
+            }
+            apply {
+                JointMeasureHelper(bases[0], q, qubits[0]);
+                JointMeasureHelper(bases[1], q, qubits[1]);
+                JointMeasureHelper(bases[2], q, qubits[2]);
+                JointMeasureHelper(bases[3], q, qubits[3]);
+                JointMeasureHelper(bases[4], q, qubits[4]);
+                JointMeasureHelper(bases[5], q, qubits[5]);
+                JointMeasureHelper(bases[6], q, qubits[6]);
+                JointMeasureHelper(bases[7], q, qubits[7]);
+            }
+            return MResetZ(q);
+        }
+        else {
+            fail "Too many qubits specified in call to Measure.";
         }
-        return res;
     }
```

Comment:
I assume there is a reason why `operation Measure()` cannot to be implemented iteratively or recursively (for any length of the params `bases : Pauli[], qubits : Qubit[]`). If that's the case then it probably makes sense to change either the name of this operation `Measure()` or the doc comments of it to indicate that `Length(bases) <= 8` is supported only.

Reply:
Yes, it has to do with avoiding variable indexing, recursion, and callable variables, which also motived the changes in #992. By avoiding those qualities, we can compile away complexity here with higher confidence for execution on current hardware.

Unfortunately, the docs must match across `Measure` implementations because they are essentially all the same function. Or at least, that's the strategy we've been employing thus far. @cgranade do you have any thoughts on documentation here? With the current way target packages are pulled in, callers using the IDE to preview documentation will always see the default QSharp.Core package docs, and I've kept them the same throughout because I didn't want to confuse any of the automatic doc generation logic with two identically named (and namespaced) operations with different docs. Any advice here?

Reply:
Let me tag in @msoeken here as well. In any case, I think we should document this restriction in QSharp.Core, noting that the limitation to eight qubits is on some targets (maybe in `/// # Remarks`?). Where we document profiles and targets, we can note the eight-qubit limitation there as well, perhaps?

Reply:
That seems like a good option. It may be something we should do across the board, since the target packages now also have a limit of 8 control qubits in the controlled operations.

Reply:
We should continue the documentation discussion, however I'm going to go ahead and merge the PR to get the code changes in.